R/teal_slice.R

```diff
@@ -1,36 +1,148 @@
+#' @param choices (optional `vector`) specifying allowed choices;
+#' When specified it should be a subset of values in variable denoted by `varname`;
+#' Type and size depends on variable type.
+#' @param selected (optional `vector`) of selected values from `choices`;
+#' Type and size depends on variable type.
+#' @param multiple (optional `logical(1)`) flag specifying whether more than one value can be selected;
+#' only applicable to `ChoicesFilterState` and `LogicalFilterState`
+#' @param keep_na (optional `logical(1)` or `NULL`) flag specifying whether to keep missing values
+#' @param keep_inf (optional `logical(1)` or `NULL`) flag specifying whether to keep infinite values
+#' @param fixed (`logical(1)`) flag specifying whether to fix this filter state (forbid setting state)
+#' @param locked (`logical(1)`) flag specifying whether to lock this filter state (forbid disabling and removing)
+#' @param title (optional `character(1)`) title of the filter. Ignored when `varname` is set.
+#' @param ... additional arguments which can be handled by extensions of `teal.slice` classes.
+#'
+#' @return `teal.slice` object
 #'
-#' `filter_var` creates a `teal_slice` object, which specifies a filter for a single variable,
-#' passed to and resolved by `FilterState` objects.
-#' `filter_settings` collates multiple `teal_slice` objects into `teal_slices`,
+#' @examples
+#' teal_slice(
+#'   dataname = "data",
+#'   id = "Female adults",
+#'   expr = "SEX == 'F' & AGE >= 18",
+#'   title = "Female adults"
+#' )
+#' teal_slice(
+#'  dataname = "data",
+#'  varname = "var",
+#'  choices = c("F", "M", "U"),
+#'  selected = "F",
+#'  keep_na = TRUE,
+#'  keep_inf = TRUE,
+#'  fixed = FALSE,
+#'  locked = FALSE,
+#'  multiple = TRUE,
+#'  id = "Gender",
+#'  extra_arg = "extra"
+#' )
+#' @export
+teal_slice <- function(dataname,
+                        varname,
+                        id,
+                        expr,
+                        choices = NULL,
+                        selected = NULL,
+                        keep_na = NULL,
+                        keep_inf = NULL,
+                        fixed = if (!missing(expr)) TRUE else FALSE,
+                        locked = FALSE,
+                        multiple = TRUE,
+                        title = NULL,
+                        ...) {
+  checkmate::assert_string(dataname)
+  checkmate::assert_flag(fixed)
+  checkmate::assert_flag(locked)
+
+  formal_args <- as.list(environment())
+  if (!missing(expr) && !missing(varname)) {
+    stop("Must provide either `expr` or `varname`.")
+  } else if (!missing(expr)) {
+    ts_expr_args <- c("dataname", "id", "expr", "fixed", "locked", "title")
+    formal_args <- formal_args[ts_expr_args]
+    checkmate::assert_string(id)
+    checkmate::assert_string(title)
+    checkmate::assert_string(expr)
+    ans <- do.call(shiny::reactiveValues, c(formal_args, list(...)))
+    class(ans) <- c("teal_slice_expr", "teal_slice", class(ans))
+    ans
```

Comment:
Perhaps remove it from the conditional and place it at the bottom of the function?